velox/common/memory/ArbitrationOperation.h

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/common/base/Counters.h"
+#include "velox/common/base/GTestMacros.h"
+#include "velox/common/base/StatsReporter.h"
+#include "velox/common/future/VeloxPromise.h"
+#include "velox/common/memory/ArbitrationParticipant.h"
+#include "velox/common/memory/Memory.h"
+
+namespace facebook::velox::memory {
+
+/// Manages the execution of a memory arbitration request within the arbitrator.
+class ArbitrationOperation {
+ public:
+  ArbitrationOperation(
+      ScopedArbitrationParticipant&& pool,
+      uint64_t requestBytes,
+      uint64_t timeoutMs);
+
+  ~ArbitrationOperation();
+
+  enum class State {
+    kInit,
+    kWaiting,
+    kRunning,
+    kFinished,
+  };
+  State state() const {
+    return state_;
+  }
+  static std::string stateName(State state);
+
+  /// Returns the corresponding arbitration participant.
+  const ScopedArbitrationParticipant& participant() {
+    return participant_;
+  }
+
+  /// Invoked to start arbitration execution on the arbitration participant. The
+  /// latter ensures the serialized execution of arbitration operations from the
+  /// same query with one at a time. So this method blocks until all the prior
+  /// arbitration operations finish.
+  void start();
```

Comment:
similarly, start() method does not technically start the arbitration. Shall we rename it?

Reply:
Replied